zcash_client_backend/src/proto.rs

```diff
@@ -183,3 +208,300 @@ impl service::TreeState {
         read_commitment_tree::<Node, _, NOTE_COMMITMENT_TREE_DEPTH>(&sapling_tree_bytes[..])
     }
 }
+
+/// Constant for the V1 proposal serialization version.
+pub const PROPOSAL_SER_V1: u32 = 1;
+
+/// Errors that can occur in the process of decoding a [`Proposal`] from its protobuf
+/// representation.
+#[derive(Debug, Clone)]
+pub enum ProposalDecodingError<DbError> {
+    /// The ZIP 321 transaction request URI was invalid.
+    Zip321(Zip321Error),
+    /// A transaction identifier string did not decode to a valid transaction ID.
+    TxIdInvalid(TryFromSliceError),
+    /// A failure occurred trying to retrievean unspent note or UTXO from the wallet database.
+    InputRetrieval(DbError),
+    /// The unspent note or UTXO corresponding to a proposal input was not found in the wallet
+    /// database.
+    InputNotFound(TxId, PoolType, u32),
+    /// The transaction balance, or a component thereof, failed to decode correctly.
+    BalanceInvalid,
+    /// Failed to decode a ZIP-302 compliant memo from the provided memo bytes.
+    MemoInvalid(memo::Error),
+    /// The serialization version returned by the protobuf was not recognized.
+    VersionInvalid(u32),
+    /// The proposal did not correctly specify a standard fee rule.
+    FeeRuleNotSpecified,
+    /// The proposal violated balance or structural constraints.
+    ProposalInvalid(ProposalError),
+    /// A `sapling_inputs` field was present, but contained no input note references.
+    EmptySaplingInputs,
```

Comment:
Should this be `EmptyShieldedInputs(PoolType)`? Non-blocking.

Reply:
Hm, yeah, Orchard will probably have the same issue, so I should generalize it now.